WireUI/Sources/WireUIFoundation/Main/MainCoordinator.swift

```diff
@@ -0,0 +1,364 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import UIKit
+import WireFoundation
+
+// swiftlint:disable opening_brace
+
+/// Manages the main navigation and the layout changes of the application after a successful login.
+///
+/// The MainCoordinator class is the central controller for the app's navigation and layout management.
+/// It receives references to ``MainTabBarControllerProtocol`` and ``MainSplitViewControllerProtocol``
+/// conforming instances and is responsible for managing transitions between different split layout states (collapsed and expanded)
+/// as well as handling navigation logic.
+///
+/// Both, the split view controller as well as the tab view controller actually install `UINavigationController`
+/// instances and then put or remove the content view controllers into/from `viewControllers` array.
+
+@MainActor
+public final class MainCoordinator<
+
+    SplitViewController: MainSplitViewControllerProtocol,
+    TabBarController: MainTabBarControllerProtocol,
+    NewConversationBuilder: MainCoordinatorInjectingViewControllerBuilder,
+    SelfProfileBuilder: MainCoordinatorInjectingViewControllerBuilder
+
+>: NSObject, MainCoordinatorProtocol, UISplitViewControllerDelegate, UITabBarControllerDelegate where
+
+    SplitViewController.Sidebar: MainSidebarProtocol,
+    SplitViewController.ConversationList == TabBarController.ConversationList,
+    SplitViewController.Archive == TabBarController.Archive,
+    SplitViewController.Settings == TabBarController.Settings
+{
+    // swiftlint:enable opening_brace
+
+    // MARK: - Private Properties
+
+    private weak var splitViewController: SplitViewController!
+    private weak var tabBarController: TabBarController!
```

Comment:
is there a reason to make them weak ?  having them weak + forced unwrapped feels a bit counter intuitive, also wondering if they need be released from memory at any point.

Reply:
I'm not 100% sure if it's required but it shouldn't hurt neither. 
What I want to avoid is a retain cycle.
Since we pass the main coordinator into content view controllers (like the conversation list), which themselves a retained by their container view controllers (split/tabbar), if we also keep a strong reference from the main coordinator to the container view controllers, won't it be a cycle?